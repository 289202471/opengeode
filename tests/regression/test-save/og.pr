SYSTEM og;
    /* CIF TEXT (159, 221), (289, 188) */
    -- Text area for declarations and comments

use dv comment 'dataview.asn';

signal run;
signal saved_signal(BoolType);
signal we;
    /* CIF ENDTEXT */
    CHANNEL c
        FROM ENV TO og WITH run,
saved_signal;
        FROM og TO ENV WITH we;
    ENDCHANNEL;
    BLOCK og;
        SIGNALROUTE r
            FROM ENV TO og WITH run,
saved_signal;
            FROM og TO ENV WITH we;
        CONNECT c AND r;
        /* CIF PROCESS (225, 50), (150, 75) */
        PROCESS og;
            /* CIF TEXT (335, 58), (396, 136) */
            --  Demonstrate how to simulate the behaviour of the SAVE
-- signal. The actual SAVE implies dynamic memory allocation.
-- But using a controlled-size array and a continuous signal it
-- is easy to reproduce the same behaviour.

dcl save_buffer SeqOf;

dcl param BoolType;
            /* CIF ENDTEXT */
            /* CIF START (0, 226), (70, 35) */
            START;
                /* CIF NEXTSTATE (0, 276), (70, 35) */
                NEXTSTATE wait;
            /* CIF STATE (291, 227), (70, 35) */
            STATE wait;
                /* CIF INPUT (91, 282), (84, 35) */
                INPUT run;
                    /* CIF NEXTSTATE (79, 332), (107, 35) */
                    NEXTSTATE Running;
                /* CIF INPUT (296, 282), (162, 35) */
                INPUT saved_signal(param);
                    /* CIF DECISION (296, 332), (161, 50) */
                    DECISION length(save_buffer);
                            /* CIF ANSWER (253, 402), (60, 34) */
                            (<10):
                                /* CIF TASK (195, 451), (174, 40) */
                                TASK save_buffer :=
save_buffer // {param};
                            /* CIF ANSWER (457, 402), (53, 34) */
                            (10):
                                /* CIF PROCEDURECALL (382, 451), (202, 38) */
                                CALL writeln('Buffer Overflow!');
                    ENDDECISION;
                    /* CIF NEXTSTATE (342, 506), (70, 35) */
                    NEXTSTATE wait;
            ENDSTATE;
            /* CIF STATE (929, 221), (88, 35) */
            STATE Running;
                /* CIF INPUT (1053, 276), (70, 35) */
                INPUT Run;
                    /* CIF PROCEDURECALL (990, 326), (194, 35) */
                    CALL writeln('Already running');
                    /* CIF NEXTSTATE (1053, 376), (70, 35) */
                    NEXTSTATE -;
                /* CIF PROVIDED (718, 276), (181, 35) */
                PROVIDED length(save_buffer) > 0;
                    /* CIF TASK (717, 326), (182, 35) */
                    TASK param := save_buffer(0);
                    /* CIF TASK (636, 376), (344, 35) */
                    TASK save_buffer := save_buffer(1, length(save_buffer));
                    /* CIF TASK (647, 426), (323, 56) */
                    TASK save_buffer := if length(save_buffer) > 1
then {true} else save_buffer // {false } fi;
                    /* CIF TASK (650, 497), (316, 56) */
                    TASK save_buffer := if length(save_buffer) > 1
then save_buffer(1, length(save_buffer))
else {} fi;
                    /* CIF PROCEDURECALL (708, 568), (200, 35) */
                    CALL writeln('Running: ', param);
                    /* CIF NEXTSTATE (764, 618), (88, 35) */
                    NEXTSTATE Running;
            ENDSTATE;
        ENDPROCESS og;
    ENDBLOCK;
ENDSYSTEM;