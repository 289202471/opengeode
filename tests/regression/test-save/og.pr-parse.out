[INFO] Starting to parse model: test-save/og.pr
[ERROR] Failed to parse model.
    line 3:4 no viable alternative at input '/* CIF TEXT (159, 221), (289, 188) */\n    -- Text area for declarations and comments\n\nuse dv comment 'dataview.asn';\n\nsignal run;\nsignal saved_signal(BoolType);\nsignal we;\n    /* CIF ENDTEXT */\n    CHANNEL c\n        FROM ENV TO og WITH run,\nsaved_signal;\n        FROM og TO ENV WITH we;\n    ENDCHANNEL;\n    BLOCK og;\n        SIGNALROUTE r\n            FROM ENV TO og WITH run,\nsaved_signal;\n            FROM og TO ENV WITH we;\n        CONNECT c AND r;\n        /* CIF PROCESS (225, 50), (150, 75) */\n        PROCESS og;\n            /* CIF TEXT (150, 75), (396, 136) */\n            --  Demonstrate how to simulate the behaviour of the SAVE\n-- signal. The actual SAVE implies dynamic memory allocation.\n-- But using a controlled-size array and a continuous signal it\n-- is easy to reproduce the same behaviour.\n\ndcl save_buffer SeqOf;\n\ndcl param BoolType;\n            /* CIF ENDTEXT */\n            /* CIF PROCEDURE (0, 106), (120, 35) */\n            PROCEDURE SAVE_PARAM;\n                /* CIF TEXT (1, 27), (289, 140) */\n                fpar in param BoolType;\n                /* CIF ENDTEXT */\n                /* CIF START (146, 192), (70, 28) */\n                START;\n                    /* CIF DECISION (100, 238), (161, 50) */\n                    DECISION length(save_buffer);\n                            /* CIF ANSWER (58, 308), (60, 34) */\n                            (<10):\n                                /* CIF TASK (0, 357), (177, 50) */\n                                TASK save_buffer :=\nsave_buffer // {param};\n                            /* CIF ANSWER (262, 308), (53, 34) */\n                            (10):\n                                /* CIF PROCEDURECALL (187, 357), (202, 38) */\n                                CALL writeln('Buffer Overflow!');\n                    ENDDECISION;\n                    /* CIF RETURN (163, 422), (35, 35) */\n                    RETURN ;\n            ENDPROCEDURE;\n            /* CIF PROCEDURE (5, 163), (113, 35) */\n            PROCEDURE GET_PARAM;\n                /* CIF TEXT (341, 385), (289, 140) */\n                -- Text area for declarations and comments\n\ndcl idx myinteger := 0;\n                /* CIF ENDTEXT */\n                /* CIF START (153, 81), (70, 35) */\n                START;\n                    /* CIF TASK (94, 131), (188, 35) */\n                    TASK param := save_buffer(0);\n                    /* CIF TASK (80, 181), (215, 35) */\n                    TASK idx := length (save_buffer) - 1;\n                    /* CIF TASK (46, 231), (282, 56) */\n                    TASK save_buffer := if length(save_buffer) > 1\nthen save_buffer(1, idx) else {} fi;\n                    /* CIF RETURN (170, 302), (35, 35) */\n                    RETURN ;\n            ENDPROCEDURE;\n            /* CIF START (97, 224), (70, 35) */\n            START;\n                /* CIF NEXTSTATE (97, 274), (70, 35) */\n                NEXTSTATE wait;\n            /* CIF STATE (479, 236), (88, 35) */\n            STATE Running;\n                /* CIF INPUT (588, 291), (70, 35) */\n                INPUT Run;\n                    /* CIF PROCEDURECALL (525, 341), (194, 35) */\n                    CALL writeln('Already running');\n                    /* CIF NEXTSTATE (588, 391), (70, 35) */\n                    NEXTSTATE -;\n                /* CIF PROVIDED (324, 291), (181, 35) */\n                PROVIDED length(save_buffer) > 0;\n                    /* CIF PROCEDURECALL (358, 341), (113, 35) */\n                    CALL GET_PARAM;\n                    /* CIF PROCEDURECALL (314, 391), (200, 35) */\n                    CALL writeln('Running: ', param);\n                    /* CIF NEXTSTATE (370, 441), (88, 35) */\n                    NEXTSTATE Running;\n            ENDSTATE;\n            /* CIF STATE (97, 274), (70, 35) */\n            STATE wait;\n                /* CIF INPUT (7, 329), (84, 35) */\n                INPUT run;\n                    /* CIF NEXTSTATE (5, 379), (88, 35) */\n                    NEXTSTATE Running;\n                /* CIF INPUT (103, 329), (162, 35) */\n                INPUT saved_signal(param);\n                    /* CIF PROCEDURECALL (100, 379), (168, 35) */\n                    CALL SAVE_PARAM(param);\n                    /* CIF NEXTSTATE (149, 429), (70, 35) */\n                    NEXTSTATE wait;\n            ENDSTATE;\n        ENDPROCESS og;\n    ENDBLOCK;\n'
